perfkitbenchmarker/alicloud/util.py

```diff
@@ -0,0 +1,136 @@
+
+ALI_PREFIX = ['aliyuncli']
+ROOT = 'root'
+FLAGS = flags.FLAGS
+
+
+REGION_HZ = 'cn-hangzhou'
+
+
+def GetEncodedCmd(cmd):
+  cmd_line = ' '.join(cmd)
+  cmd_args = shlex.split(cmd_line)
+  return cmd_args
+
+
+def GetRegionByZone(zone):
+  if zone.find(REGION_HZ) != -1:
+      return REGION_HZ
+  s = zone.split('-')
+  if s[0] == 'cn':
+    s.pop()
+    return '-'.join(s)
+  else:
+    return zone[:-1]
+
+
+def AddTags(resource_id, resource_type, region, **kwargs):
+  """Adds tags to an AliCloud resource created by PerfKitBenchmarker.
+
+  Args:
+    resource_id: An extant AliCloud resource to operate on.
+    resource_type: The type of the resource.
+    region: The AliCloud region 'resource_id' was created in.
+    **kwargs: dict. Key-value pairs to set on the instance.
+  """
+  if not kwargs:
+    return
+
+  tag_cmd = ALI_PREFIX + [
+      'ecs',
+      'AddTags',
+      '--RegionId %s' % region,
+      '--ResourceId %s' % resource_id,
+      '--ResourceType %s' % resource_type]
+  for index, (key, value) in enumerate(kwargs.iteritems()):
+    tag_cmd.append('--Tag{0}Key {1} --Tag{0}Value {2}'
+                   .format(index + 1, key, value))
+  tag_cmd = GetEncodedCmd(tag_cmd)
+  vm_util.IssueRetryableCommand(tag_cmd)
+
+
+def AddDefaultTags(resource_id, resource_type, region):
+  """Adds tags to an AliCloud resource created by PerfKitBenchmarker.
+
+  By default, resources are tagged with "owner" and "perfkitbenchmarker-run"
+  key-value
+  pairs.
+
+  Args:
+    resource_id: An extant AliCloud resource to operate on.
+    resource_type: The type of the 'resource_id'
+    region: The AliCloud region 'resource_id' was created in.
+  """
+  tags = {'owner': FLAGS.owner, 'perfkitbenchmarker-run': FLAGS.run_uri}
+  AddTags(resource_id, resource_type, region, **tags)
+
+
+@vm_util.Retry()
+def AddPubKeyToHost(host_ip, password, keyfile, username):
+  public_key = str()
+  if keyfile:
+    keyfile = os.path.expanduser(keyfile)
+    if not paramiko:
+      raise Exception('`paramiko` is required for pushing keyfile to ecs.')
+    with open(keyfile, 'r') as f:
+      public_key = f.read()
+  ssh = paramiko.SSHClient()
+  ssh.set_missing_host_key_policy(paramiko.AutoAddPolicy())
+  try:
+    ssh.connect(
+        host_ip,
+        username=ROOT,  # We should use root to add pubkey to host
+        password=password,
+    )
+    if username == ROOT:
+      command_list = [
+          'mkdir .ssh; echo "%s" >> ~/.ssh/authorized_keys' % public_key
+      ]
+    else:
+      command_list = [
+          'echo "{0} ALL = NOPASSWD: ALL" >> /etc/sudoers'.format(username),
+          'useradd {0} --home /home/{0} --shell /bin/bash -m'.format(username),
+          'mkdir /home/{0}/.ssh'.format(username),
+          'echo "{0}" >> /home/{1}/.ssh/authorized_keys'
+          .format(public_key, username),
+          'chown -R {0}:{0} /home/{0}/.ssh'.format(username)]
+    command = ';'.join(command_list)
+    ssh.exec_command(command)
```

Comment:
how about IssueRemoteCommand?  One b of these should work on this case.
On Nov 5, 2015 5:42 AM, "Geoffrey" <notifications@github.com> wrote:

> In perfkitbenchmarker/alicloud/util.py
> <https://github.com/GoogleCloudPlatform/PerfKitBenchmarker/pull/611#discussion_r44011894>
> :
>
> > +        password=password,
> > +    )
> > +    if username == ROOT:
> > +      command_list = [
> > +          'mkdir .ssh; echo "%s" >> ~/.ssh/authorized_keys' % public_key
> > +      ]
> > +    else:
> > +      command_list = [
> > +          'echo "{0} ALL = NOPASSWD: ALL" >> /etc/sudoers'.format(username),
> > +          'useradd {0} --home /home/{0} --shell /bin/bash -m'.format(username),
> > +          'mkdir /home/{0}/.ssh'.format(username),
> > +          'echo "{0}" >> /home/{1}/.ssh/authorized_keys'
> > +          .format(public_key, username),
> > +          'chown -R {0}:{0} /home/{0}/.ssh'.format(username)]
> > +    command = ';'.join(command_list)
> > +    ssh.exec_command(command)
>
> I failed to use IssueCommand here because it should post the password to
> remote vm.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/PerfKitBenchmarker/pull/611/files#r44011894>
> .
>